Screen ordering and colouring

SimonBrandner opened the pull request:
Things I've added:

+ `getScreenNames()` method

SimonBrandner commented:
I've added a basic implementation of the UI for screen ordering. It still needs quite some polishing.

SimonBrandner commented:
I think I've successfully implemented screen ordering. I've also reformated some QML using [`mblode.pretty-formatter`](https://marketplace.visualstudio.com/items?itemName=mblode.pretty-formatter) VS Code extension.

duvholt commented on src/i3pager.cpp:
```diff
@@ -41,7 +41,24 @@ void I3Pager::handleI3Events() {
+        auto screens = conn.get_outputs();
+
+        for (auto& screen : screens) {
+            screenList.append(QString::fromStdString(screen->name));
```
`screen` has an `active` property which we should use here to filter out inactive screens like `xroot-0`.

SimonBrandner replied:
Oh, I didn't know that. Thank you for letting me know. Should be fixed now.

duvholt commented on src/i3pager.cpp:
```diff
@@ -81,6 +98,10 @@ QList<Workspace> I3Pager::getWorkspaces(bool filterByCurrentScreen) {
         workspaceList = Workspace::filterByCurrentScreen(workspaceList, this->currentScreenPrivate);
     }
+
+    if (orderWorkspacesBy == "screen" && !screenOrder.isEmpty()) {
```
Not a huge fan of hardcoded strings like this, but I'm not sure how we could share something like an enum between C++ and QML code either.

SimonBrandner replied:
Neither am I. Accessing a `C++` enum from `QML` seems to be [possible](https://stackoverflow.com/questions/20089196/how-to-access-c-enum-from-qml). Although I have no idea how to make this work with [`KConfig`](https://api.kde.org/frameworks/kconfig/html/classKCoreConfigSkeleton.html). I am not sure if this is worth trying to implement on a codebase this small.

duvholt replied:
Yeah, it's probably not worth the effort.

duvholt commented on plasmoid/contents/ui/screens.js:
```diff
@@ -0,0 +1,31 @@
+			return true;
+		}
+	}
+	return false;
```
Looks like this file is indented with tabs. Could you convert it to spaces so it matches the other files?

SimonBrandner replied:
Sorry, I totally forgot, I've fixed it now.

duvholt commented on plasmoid/contents/ui/configScreens.qml:
```diff
@@ -0,0 +1,145 @@
+import I3Pager 1.0
+import "screens.js" as ScreensJS
+
+ColumnLayout {
```
The screen configuration tab looks nice :+1:

SimonBrandner replied:
Thanks :)

duvholt reviewed (commented):
Generally looks nice, but sometimes after changing configuration the whole pager crashes. It actually manages to crash plasmashell too which I didn't even know was possible. 

I haven't found a reliable way to reproduce the crash yet and it might not even be related to the changes in this PR.

duvholt commented:
I found a way to reproduce the crash now.
1. Open the new `Screens` tab
2. Change to any other tab
3. Plasma now crashes on the next i3 event (for example by changing focus to another window)

I think it's related to this in `configScreens.qml`:
```qml
    I3Pager {
        id : i3pager
    }
```
Maybe this  happens when unloading I3Pager without properly cleaning up the thread used to listen to i3ipc events.

SimonBrandner commented:
> I found a way to reproduce the crash now.

I can reproduce it too. I've noticed this strange behaviour once or twice before. Somehow I assumed it had something to do with `plasmawindowed`.

duvholt commented:
I think I fixed the crash in #13

SimonBrandner commented:
> I think I fixed the crash in #13

The problem seems to persist at least on my side.

SimonBrandner commented:
Ok, I need to take a break. Is this really happening only on my side of things?

duvholt commented:
Sorry, looks like I didn't test good enough. It's still crashing for me.
I still think it's related to some missing cleanup on destruction since it happens after unloading the screens tab.

Plasma is creating a stacktrace, but it's only showing compiled symbols which makes it a bit hard to see where the actual error is happening: `#5  0x00007fab0809f11c _ZN4sigc8internal12signal_emit1IvRKN5i3ipc17workspace_event_tENS_3nilEE4emitEPNS0_11signal_implES5_ (libplasmoidplugin.so + 0x4311c)`

SimonBrandner commented:
I didn't manage to figure out what exactly is causing the issue but I think I've found a [workaround](https://github.com/SimonBrandner/i3-pager/commit/4ebe350736abe0cec0364686a434bffd3de20afa). 

Do we want to keep looking for a fix or are we going to base our following actions on the workaround? If so do we want to stay with the workaround or do want to rewrite the code somehow?

I am thinking about making a separate class for events and a separate class for retrieving data from i3ipc.

duvholt commented:
I think the workaround is fine for now, but in the long term I want a more robust solution for listening to i3 events. 

I'm fairly sure the problem is I'm emitting signals in a non threadsafe way. I started rewriting the i3 listener to use a worker-object as described here: https://doc.qt.io/qt-5/qthread.html#details
Ideally it should use Qt's event listener for polling, but that wasn't as straight forward as I hoped and I got stuck while trying to use i3ipc++ event socket with Qt. I'll probably take another look tomorrow and see I find a solution or a workaround.

It might be worth replacing i3ipc++ with this project which uses Qt, but unfortunately it doesn't look like it's actively maintained. https://gitlab.com/hantz/qi3pc

Also found out that starting plasmashell with gdb works great for debugging crashes.

SimonBrandner commented:
> I think the workaround is fine for now

Ok, I'll finish implementing the screen colouring with the workaround.

Thanks for all the work trying to find a better long term solution :+1:

duvholt commented:
@SimonBrandner could you check if #14 fixes the crash for you? I ended up creating a new QThread which communicates with I3Pager using signals.

SimonBrandner commented:
> could you check if #14 fixes the crash for you?

@duvholt This fixed the crash! :+1:

SimonBrandner commented:
@duvholt I've implemented the screen colouring. It may deserve some polishing. What do you think?

SimonBrandner commented:
I've been playing with the UI and have created something quite different from the original. So I've made another [branch](https://github.com/SimonBrandner/i3-pager/tree/ui-tweaks) for it. Let me know what you think.

duvholt commented:
I've been trying out the new screen colors and I like the idea but in practice I find it hard to find fitting colors easily. 

This is probably because I'm using a semi transparent plasma panel, but I ended up liking semi transparent colors the best. 
However the visible/focused indicators end up looking out of place imo. 

Just changing screen colors to some semi transparent color:
![image](https://user-images.githubusercontent.com/1637715/97228327-0ab56680-17d7-11eb-8e37-8a1b94e62ed8.png)

Using screen color as visible indicator (it's too subtle but I just hacked it in):
![image](https://user-images.githubusercontent.com/1637715/97228218-e2c60300-17d6-11eb-842a-7d50cd6c5713.png)

Either way I think it makes sense to have some default colors that look decent enough, but still let the user tweak everything to fit their own setup. At least that's one of the things I really like about KDE Plasma.

I'm wondering if it makes sense to create some graphical mockups at this point so that it's easier to compare different setups. 
I have some experience using Figma, but if that's more hassle for you than just testing out things with QML directly I won't force you to use it :smile:

SimonBrandner commented:
Have you tried [this branch](https://github.com/SimonBrandner/i3-pager/tree/ui-tweaks)? I think I've reached something that's better than the `develop` branch but it is still imperfect.

![Screenshot_20201027_081001](https://user-images.githubusercontent.com/25768714/97268568-5b0ae380-182c-11eb-9dbe-3ce3d9bc4272.png)

> in practice I find it hard to find fitting colors easily

I basically agree. 

> This is probably because I'm using a semi transparent plasma panel, but I ended up liking semi transparent colors the best.

I am using a transparent panel as well. I think it would be the best to implement a global opacity setting.

> I have some experience using Figma, but if that's more hassle for you than just testing out things with QML directly I won't force you to use it

I've never used Figma but I can give it a try and see how it goes.

SimonBrandner commented:
Learning the basics of Figma wasn't so hard but I found out that I am almost useless when it comes to designing stuff so I asked a friend for help. This is what we ended up with:

![Screenshot_20201027_131941](https://user-images.githubusercontent.com/25768714/97300834-3e37d580-1857-11eb-9e8d-d347146a203f.png)

The first two are basically what we have now. 

We reached the conclusion that it may be a good idea to use similar colours. We also came up with the idea of adding some spacers.

I'll send you an invite (hope it worked).

SimonBrandner commented:
I've fixed some mistakes and issues. 

I've also added sharp corners to the Figma file so that we can compare them with the round corners:

![Screenshot_20201027_143649](https://user-images.githubusercontent.com/25768714/97309281-17cb6780-1862-11eb-8a5c-283aca655440.png)

I am in favour of the round corners.

SimonBrandner commented:
@duvholt  Did you get a chance to check out the UI mockups?

duvholt commented:
Hey, sorry for my absence. Life's been busy lately.

Thanks for making some mockups. I think I personally prefer underline instead of border, but again it might be because I'm so used to my current setup.  How hard would it be to make this completely customizable? I'm also wondering if maybe there some native QML or Plasma components we could use that would have a higher chance of looking good by default for all themes.
Though I'm sure it's worth it to have to maintain multiple versions of more or less the same component. 


Either way I think this PR is a nice improvement. Earlier I actually was kind of forced to use it since the plasma panel on my right monitor just disappeared mid day at work. It worked pretty well :+1:

SimonBrandner commented:
> Hey, sorry for my absence. Life's been busy lately.

It's fine :) Lately, I've had a lot of stuff for school myself. 

> How hard would it be to make this completely customizable?

Making multiple options available shouldn't be a problem at all. 

> I'm also wondering if maybe there some native QML or Plasma components we could use that would have a higher chance of looking good by default for all themes.

I have my doubts about that but I'll definitely take a look.

SimonBrandner commented:
I've added an opacity spin box to the global opacity slider and an option to change styles (underline, border). I am wondering if I should try to make the rounded corners optional? I could also make the position of the underline indicator customizable (top, bottom).

SimonBrandner commented:
I think everything is now custumizable

duvholt commented:
Neat! Just tested it out and it works great. I'll try to do a code review later today.

duvholt commented on plasmoid/contents/ui/main.qml:
```diff
@@ -8,157 +8,206 @@ import QtQuick.Window 2.13
-    Layout.fillHeight: true
-    spacing: 0
+    function colorWithAlpha(color, alpha) {
+        console.log(color.r);
```
Debug leftover I'm guessing?

SimonBrandner replied:
Yes, that was a debug leftover.

duvholt commented on plasmoid/contents/ui/screens.js:
```diff
@@ -0,0 +1,49 @@
+
+    for (var i = 0; i < screenListModel.count; i++) {
+        var name = screenListModel.get(i).screenName;
+        var color = Qt.lighter(screenListModel.get(i).screenColor, 1);
```
Does `Qt.lighter` with a factor of 1 actually do anything?

SimonBrandner replied:
This is a workaround. I couldn't find a method to create `QColor` from `string` in Javascript, so I used `Qt.lighter()` with the factor of `1` to do this. Is there a normal way to do this?

duvholt replied:
Huh, that's weird. The closest thing I can find is Qt.rgba, but that function doesn't offer the same api. 
Maybe create a helper function with a short comment about why Qt.lighter is used? Or a helper function that uses Qt.rgba.

duvholt commented on plasmoid/contents/ui/screens.js:
```diff
@@ -0,0 +1,49 @@
+                screenListModel.append({
+                    screenName: screenName,
+                    screenColor: defaultColors[screenListModel.count]
+                });
```
Since the only difference here is the `defaultColors` index I think it's easier to read if the index is first selected and then the append happens afterwards. Could be done by doing something like `defaultColorIndex = Math.min(screenListModel.count, defaultColors.length - 1)`.

SimonBrandner replied:
```js
function loadNewScreens() {
    var defaultColors = ["#004040", "#007D7D", "#00D7D7"];
    for (var screenName of i3pager.getScreenNames()) {
        if (!isScreenNameOnList(screenName, cfg_screenNameList)) {
            screenListModel.append({
                screenName: screenName,
                screenColor: defaultColors[Math.min(screenListModel.count, defaultColors.length - 1)]
            });
        }
    }
}
```
Something like this? Do I understand correctly?

duvholt replied:
Yeah, however I would create a variable for the index to reduce the amount of things happening on one line. That's a personal preference though.
```js
        if (!isScreenNameOnList(screenName, cfg_screenNameList)) {
            var colorIndex = Math.min(screenListModel.count, defaultColors.length - 1);
            screenListModel.append({
                screenName: screenName,
                screenColor: defaultColors[colorIndex]
            });
        }
```

duvholt reviewed (commented):
Just some nitpicking. Overall the code looks good to me 👍

SimonBrandner commented:
I think everything is done.

duvholt commented:
Thanks!

SimonBrandner commented:
You're welcome :)
